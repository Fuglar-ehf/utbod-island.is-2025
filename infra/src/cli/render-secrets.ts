import { generateYamlForEnv } from '../dsl/serialize-to-yaml'
import { UberChart } from '../dsl/uber-chart'
import { Envs } from '../environments'
import { charts } from '../uber-charts/all-charts'
import { SSM } from '@aws-sdk/client-ssm'

const API_INITIALIZATION_OPTIONS = {
  region: 'eu-west-1',
  maxAttempts: 10,
}

const EXCLUDED_ENVIRONMENT_NAMES = [
  'DB_PASSWORD',
  'NOVA_USERNAME',
  'NOVA_PASSWORD',
]

const client = new SSM(API_INITIALIZATION_OPTIONS)

export const renderSecretsCommand = async (service: string) => {
  renderSecrets(service).catch((error) => {
    if (error.name === 'CredentialsProviderError') {
      console.error(
        'Could not load AWS credentials from any providers. Did you forget to configure environment variables, aws profile or run `aws sso login`?',
      )
    } else {
      console.error(error)
    }
    process.exit(1)
  })
}

export const renderSecrets = async (service: string) => {
  const urls: string[] = []
  const uberChart = new UberChart(Envs.dev)
  const services = Object.values(charts).map(
    (chart) => generateYamlForEnv(uberChart, ...chart.dev).services,
  )

  const secretRequests: [string, string][] = services
    .map((svc) => {
      return Object.entries(svc)
        .map(([serviceName, config]) => {
          if (serviceName == service) {
            return Object.entries(config.secrets)
          }
          return []
        })
        .reduce((p, c) => p.concat(c), [])
    })
    .reduce((p, c) => p.concat(c), [])
    .filter(([envName]) => !EXCLUDED_ENVIRONMENT_NAMES.includes(envName))

  const values = await getParams(secretRequests.map(([_, ssmName]) => ssmName))

  secretRequests.forEach(([envName, ssmName]) => {
    console.log(`export ${envName}=${values[ssmName]}`)
  })
}

const getParams = async (
  ssmNames: string[],
): Promise<{ [name: string]: string }> => {
  const chunks = ssmNames.reduce((all: string[][], one: string, i: number) => {
    const ch = Math.floor(i / 10)
    all[ch] = ([] as string[]).concat(all[ch] || [], one)
    return all
  }, [])

  const allParams = await Promise.all(
    chunks.map((Names) =>
      client.getParameters({ Names, WithDecryption: true }),
    ),
  )
  return allParams
    .map(({ Parameters }) =>
      Object.fromEntries(Parameters!.map((p) => [p.Name, p.Value])),
    )
    .reduce((p, c) => ({ ...p, ...c }), {})
}
