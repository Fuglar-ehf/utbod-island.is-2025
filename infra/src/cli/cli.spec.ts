import {
  GetParametersCommandInput,
  GetParametersCommandOutput,
  SSM,
} from '@aws-sdk/client-ssm'
import { renderSecretsCommand } from './render-secrets'

// jest.mock('./render-secrets')
// jest.mock('@aws-sdk/client-ssm', () => ({
//   SSM: jest.fn().mockImplementation(() => ({
//     getParameters: jest.fn(() => ({
//       Parameters: [
//         {
//           Name: '/k8s/api/A',
//           Value: 'A',
//         },
//         {
//           Name: 'A',
//           Value: 'B',
//         },
//         {
//           Name: '/k8s/api/HOUSING_BENEFIT_CALCULATOR_PASSWORD',
//           Value: '<<< NOT_UNDEFINED >>>',
//         },
//         {
//           Name: 'HOUSING_BENEFIT_CALCULATOR_PASSWORD',
//           Value: '<<< NOT_UNDEFINED >>>',
//         },
//       ],
//     })),
//   })),
// }))

// Mock the renderSecretsCommand function

describe('render-secrets command', () => {
  beforeEach(() => {
    jest.clearAllMocks()

    /*
     * // This is how we use ssm
     * const allParams = await Promise.all(
     *   chunks.map((Names) =>
     *     client.getParameters({ Names, WithDecryption: true }),
     *   ),
     * )
     */
  })

  it('should fetch 0 secrets when there are none', async () => {
    // Arrange
    // ;(renderSecretsCommand as jest.Mock).mockResolvedValueOnce({
    //   A: 'B',
    // }) // Assuming the function returns an array of secrets
    const argv = { service: 'emptyService' }

    // Act
    const envMap = await renderSecretsCommand(argv.service)

    // Assert
    // expect(renderSecretsCommand).toHaveBeenCalledWith('emptyService')
    // expect(renderSecretsCommand).toReturnWith({})
    expect(envMap).toStrictEqual({ A: 'B' })
  })

  it.only('should fetch all secrets when there are some', async () => {
    // Arrange
    const argv = { service: 'api' }

    // Act
    const envMap = await renderSecretsCommand(argv.service)

    // Assert
    // expect(renderSecretsCommand).toHaveBeenCalledWith('emptyService')
    // expect(renderSecretsCommand).toReturnWith({})
    expect(envMap).toStrictEqual({ A: 'B' })
  })

  /*
  it('should throw an error when the requested service does not exist', async () => {
    // Arrange
    const errorMessage = 'Service not found'
    ;(renderSecretsCommand as jest.Mock).mockRejectedValueOnce(
      new Error(errorMessage),
    )
    const argv = { service: 'nonExistentService' }

    // Act & Assert
    await expect(
      yargs
        .command(
          'render-secrets',
          'Render secrets needed by service',
          (yargs) => {
            return yargs.option('service', { demandOption: true })
          },
          async (argv) => {
            await renderSecretsCommand(argv.service)
          },
        )
        .parse(['render-secrets', '--service', argv.service]),
    ).rejects.toThrow(errorMessage)
  })

  it('should correctly fetch the correct number of secrets for a valid service', async () => {
    // Arrange
    const secrets = ['secret1', 'secret2', 'secret3']
    renderSecretsCommand.mockResolvedValueOnce(secrets)
    const argv = { service: 'validService' }

    // Act
    await yargs
      .command(
        'render-secrets',
        'Render secrets needed by service',
        (yargs) => {
          return yargs.option('service', { demandOption: true })
        },
        async (argv) => {
          await renderSecretsCommand(argv.service)
        },
      )
      .parse(['render-secrets', '--service', argv.service])

    // Assert
    expect(renderSecretsCommand).toHaveBeenCalledWith('validService')
    expect(renderSecretsCommand).toReturnWith(secrets)
  })
  */
})
