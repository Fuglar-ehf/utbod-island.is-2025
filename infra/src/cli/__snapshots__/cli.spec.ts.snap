// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`infra CLI render-local-env should map all URIs to localhost 1`] = `
Object {
  "mocks": "docker run --name mountebank -it --rm -p 2525:2525 -v /var/home/person/islandis/island.is/infra/mountebank-imposter-config.json:/app/default.json:z docker.io/bbyars/mountebank:2.8.1 start --configfile=/app/default.json",
  "services": Object {
    "my-service": Object {
      "commands": Array [
        "cd \\".\\"",
        ". ./.env.my-service",
        "echo \\"Starting my-service in $PWD\\"",
        "yarn nx serve my-service",
      ],
      "env": Object {
        "IDENTITY_SERVER_ISSUER_URL": "https://identity-server.dev01.devland.is",
        "PROD_MODE": "true",
        "SERVERSIDE_FEATURES_ON": "",
      },
    },
  },
}
`;
